File: src/main/java/dev/gustavpersson/thorincompanies/data_access_layer/repositories/CompanyRepository.java
package dev.gustavpersson.thorincompanies.data_access_layer.repositories;

import com.j256.ormlite.dao.Dao;
import dev.gustavpersson.thorincompanies.ThorinCompanies;
import dev.gustavpersson.thorincompanies.business_logic_layer.models.Company;
import dev.gustavpersson.thorincompanies.data_access_layer.Database;
import dev.gustavpersson.thorincompanies.data_access_layer.entities.CompanyEntity;

import java.sql.PreparedStatement;
import java.sql.SQLException;

public class CompanyRepository {

    private final Dao companyDao;
    private final Database database;

    public CompanyRepository(ThorinCompanies plugin) throws SQLException {

        database = new Database(plugin);
        companyDao = database.getDao(CompanyEntity.class);
    }

    public void createCompany(Company company) throws Exception {
        PreparedStatement query = database.getConnection().prepareStatement(
                "INSERT INTO companies (name, ownerUuid, createdAt) VALUES (?, ?, ?)");
        query.setString(1, company.getName());
        query.setString(2, company.getOwnerUuid());
        query.setDate(3, company.getCreatedAt());
        query.execute();
        //companyDao.create()
    }

}

File: src/main/java/dev/gustavpersson/thorincompanies/data_access_layer/repositories/CompanyRepository.kt
package dev.gustavpersson.thorincompanies.data_access_layer.repositories

import dev.gustavpersson.thorincompanies.business_logic_layer.models.NewCompany
import dev.gustavpersson.thorincompanies.business_logic_layer.models.RenameCompanyRequest
import dev.gustavpersson.thorincompanies.data_access_layer.entities.CompaniesTable
import dev.gustavpersson.thorincompanies.data_access_layer.entities.CompanyEntity
import org.jetbrains.exposed.sql.transactions.transaction
import java.time.LocalDate
import java.util.UUID

class CompanyRepository {
    fun create(newCompany: NewCompany): CompanyEntity {
        return transaction {
            CompanyEntity.new {
                this.name = newCompany.name
                this.founderUUID = newCompany.founderUUID.toString()
                this.startupCapital = newCompany.startupCapital
                this.createdAt = LocalDate.now()
            }
        }
    }

    fun update(request: RenameCompanyRequest): CompanyEntity {
        return transaction {
            val entity = CompanyEntity.findById(request.id) ?: throw Exception("Company not found")
            entity.name = request.name
            entity
        }
    }

    fun delete(id: Int) {
        transaction {
            CompanyEntity.findById(id)?.delete() ?: throw Exception("Company not found")
        }
    }

    fun findById(id: Int): CompanyEntity? {
        return transaction {
            CompanyEntity.findById(id)
        }
    }

    fun findByName(name: String): CompanyEntity? {
        return transaction {
            CompanyEntity.find { CompaniesTable.name eq name }.singleOrNull()
        }
    }

    fun findByFounder(founderUUID: UUID): List<CompanyEntity> {
        return transaction {
            CompanyEntity.find { CompaniesTable.founderUUID eq founderUUID.toString() }.toList()
        }
    }

    fun findAll(): List<CompanyEntity> {
        return transaction {
            CompanyEntity.all().toList()
        }
    }
}
